### PR TITLE
refactor: access to checkoutpages handled by guard

### PR DESCRIPTION
**Pull request by paramaeder**
## PR Type

[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[X ] Refactoring (no functional changes, no API changes)
[ ] Build-related changes
[ ] CI-related changes
[ ] Documentation content changes
[ ] Application / infrastructure changes
[ ] Other: 

## What Is the Current Behavior?
Currently you can navigate to the checkout with a non-existent shopping cart via the address bar of the browser

## What Is the New Behavior?
Now the route is protected by a guard. Navigate to /checkout or any related childRoute without an existing basket will redirect you to the root url

## Does this PR Introduce a Breaking Change?

[ ] Yes
[X ] No

### CODE REVIEW

File: src/app/pages/checkout/checkout.guard.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+@Injectable({
+  providedIn: 'root',
+})
+export class CheckoutGuard implements CanActivate {
+  constructor(private checkoutFacade: CheckoutFacade, private router: Router) {}
+
+  canActivate(): boolean | UrlTree {
+    let hasBasket: boolean;
+
+    this.checkoutFacade.basket$.subscribe(value => {
```
Comment (MaxKless) [inline src/app/pages/checkout/checkout.guard.ts:15]: I have to admit I'm a little surprised this works.  
Subscribing and then comparing the value extracted is not guaranteed to work - imagine if the `basket$` emission is a little delayed for whatever reason. In that case `hasBasket` will not be set correctly even though there might actually be a basket.

In addition, triggering a subscribe on method call without any logic to end the subscription is dangerous. Imagine if I tried to access the checkout route over and over again - every time `canActivate` is called, a new subscription will be opened. There are ways to fix this - check out an example like the [`feature-toggle.directive.ts`](https://github.com/intershop/intershop-pwa/blob/develop/src/app/core/directives/feature-toggle.directive.ts) to see how open subscriptions are canceled. However, this should only really be done in edge cases where nothing else is possible.
[Don't listen to me, listen to the rxjs team lead.](https://benlesh.medium.com/rxjs-dont-unsubscribe-6753ed4fda87).

Thankfully, there is a super simple solution here: [guards can return Observables](https://angular.io/api/router/CanActivate). So just return `this.checkoutFacade.basket$.pipe(....)` and do the value comparison in there.
Angular will take care of subscribing to the observable and using the emitted value for routing - nothing required on your part.

Also, I'm not sure if you need to return `this.router.parseUrl('/')` or if returning `false` would be sufficient.... Maybe you can check.

**Review by MaxKless — CHANGES_REQUESTED** [review]

File: src/app/pages/checkout/checkout-page.guard.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+@Injectable({
+  providedIn: 'root',
+})
+export class CheckoutPageGuard implements CanActivate {
+  constructor(private checkoutFacade: CheckoutFacade, private router: Router) {}
+
+  canActivate(): Observable<boolean | UrlTree> {
+    return this.checkoutFacade.basket$.pipe(
+      debounceTime(500),
```
Comment (Eisie96) [inline src/app/pages/checkout/checkout-page.guard.ts:16]: Why is the debounceTime needed? I think all possible scenearios, that a basket is loaded or is currently loading, are checked in the following switchMap. And this rxjs operator leads to the problem, that the checkout route is not activated for the first 500ms, although a basket could be available.

**Review by SGrueber — CHANGES_REQUESTED** [review]

File: src/app/pages/checkout/checkout-page.guard.ts (hunk 2 above)
Comment (paramaeder) [reply src/app/pages/checkout/checkout-page.guard.ts]: No debounceTime needed anymore. Guard now handles the request with a race operator

File: src/app/pages/checkout/checkout.guard.ts (hunk 1 above)
Comment (paramaeder) [reply src/app/pages/checkout/checkout.guard.ts]: Thanks Max! 
With increasing knowledge the original implementation was now completely overhauled

**Review by Eisie96 — APPROVED** [review]

**Review by MaxKless — APPROVED** [review]

**Review by MaxKless — APPROVED** [review]

**Pull request merged**